4.preprocess_features/scripts/0.convert_cytotable.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Convert SQLite output(s) to parquet files with CytoTable
+
+# ## Import libraries
+
+# In[1]:
+
+
+import pathlib
+import pandas as pd
+
+# cytotable will merge objects from SQLite file into single cells and save as parquet file
+from cytotable import convert, presets
+
+import logging
+
+# Set the logging level to a higher level to avoid outputting unnecessary errors from config file in convert function
+logging.getLogger().setLevel(logging.ERROR)
+
+
+# ## Set paths and variables
+
+# In[2]:
+
+
+# preset configurations based on typical CellProfiler outputs
+preset = "cellprofiler_sqlite_pycytominer"
+
+# update preset to include site metadata and cell counts
+joins = presets.config["cellprofiler_sqlite_pycytominer"]["CONFIG_JOINS"].replace(
+    "Image_Metadata_Well,",
+    "Image_Metadata_Well, Image_Metadata_Site, Image_Metadata_ZSlice, Image_Count_Cells,",
+)
+
+# type of file output from cytotable (currently only parquet)
+dest_datatype = "parquet"
+
+# set path to directory with SQLite files
+sqlite_dir = pathlib.Path("../3.feature_extraction/sqlite_output")
+
+# directory for processed data
+output_dir = pathlib.Path("data")
+output_dir.mkdir(parents=True, exist_ok=True)
+
+plate_names = []
+
+for file_path in sqlite_dir.iterdir():
+    plate_names.append(file_path.stem)
+
+# print the plate names and how many plates there are (confirmation)
+print(f"There are {len(plate_names)} plate(s) in this dataset. Below are the names:")
+for name in plate_names:
+    print(name)
+
+
+# ## Convert SQLite to parquet file(s) for single-cell profiles
+
+# In[3]:
+
+
+for file_path in sqlite_dir.iterdir():
+    output_path = pathlib.Path(
+        f"{output_dir}/converted_profiles/{file_path.stem}_sc_converted.parquet"
+    )
+    print("Starting conversion with cytotable for plate:", file_path.stem)
+    # Merge single cells and output as parquet file
+    convert(
+        source_path=str(file_path),
+        dest_path=str(output_path),
+        dest_datatype=dest_datatype,
+        preset=preset,
+        joins=joins,
+        chunk_size=500
+    )
+
+print("All plates have been converted with cytotable!")
+
+
+# # Load in converted profiles to update
+
+# In[4]:
+
+
+# Directory with converted profiles
+converted_dir = pathlib.Path(f"{output_dir}/converted_profiles")
+
+for file_path in converted_dir.iterdir():
+    # Load the DataFrame from the Parquet file
+    df = pd.read_parquet(file_path)
+
+    # If any, drop rows where "Metadata_ImageNumber" is NaN (artifact of cytotable)
+    df = df.dropna(subset=["Metadata_ImageNumber"])
```

Comment:
Correct, this is an old safety net from a while ago where the joins was leading to rows of NaN data which could easily be removed if you detect NaN in the ImageNumber column. I could probably remove it but I like to be safer rather than sorry haha.